### PR TITLE
[HLSL][DirectX] Move handling of resource element types into the frontend

### LINKED ISSUE #70686: [DirectX][HLSL] Remove `parseSourceType` from DXILResource
Currently, CGHLSLRuntime/HLSLResource insert a stringify'd type into metadata for the backend to interpret as a resource type and fill out fields such as ROV, Shape, and element type. These need to be handled in the frontend and instead pass the interpreted values to the backend in the metadata.

AC:
Metadata in LLVM internals doesn't need any parsing to translate into meaningful DXIL constructs
All resource attributes are handled natively within LLVM

### PR DESCRIPTION
**Pull request by bogner**
Rather than shepherding a type name all the way to the backend as a
string and attempting to parse it, get the element type out of the AST
and store that in the resource annotation metadata directly.

### CODE REVIEW

File: clang/lib/CodeGen/CGHLSLRuntime.cpp [hunk 1]
```diff
@@ -216,15 +214,62 @@ void CGHLSLRuntime::addBufferResourceAnnotation(llvm::GlobalVariable *GV,
...
+  const auto *TST = ResourceTy->getAs<TemplateSpecializationType>();
+  assert(TST && "Resource types must be template specializations");
+  ArrayRef<TemplateArgument> Args = TST->template_arguments();
+  assert(!Args.empty() && "Resource has no element type");
+
+  // At this point we have a resource with an element type, so we can assume
+  // that it's valid or we would have diagnosed the error earlier.
+  QualType ElTy = Args[0].getAsType();
+
+  // We should either have a basic type or a vector of a basic type.
```
Comment (dmpots) [inline clang/lib/CodeGen/CGHLSLRuntime.cpp:240]: What about matrix types? Do we need to handle them here as well?

File: llvm/lib/Target/DirectX/DXILResource.cpp [hunk 2]
```diff
@@ -253,35 +246,6 @@ void UAVResource::print(raw_ostream &OS) const {
   ResourceBase::print(OS, "U", "u");
 }
 
-// FIXME: Capture this in HLSL source. I would go do this right now, but I want
-// to get this in first so that I can make sure to capture all the extra
-// information we need to remove the source type string from here (See issue:
-// https://github.com/llvm/llvm-project/issues/57991).
```
Comment (dmpots) [inline llvm/lib/Target/DirectX/DXILResource.cpp:259]: Does this PR fix this issue? I did not see it was linked to this PR, but I may not be looking in the right place.

**Review by dmpots — APPROVED** [review]

**Review by python3kgae — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXILResource.cpp (hunk 2 above)
Comment (bogner) [reply llvm/lib/Target/DirectX/DXILResource.cpp:259]: This comment was kind of stale. That issue only mentioned capturing shape and ROV, which are both captured in the HLSL source before this change (and indeed the issue is closed). The issue this PR fixes is the obvious extension of that one, https://github.com/llvm/llvm-project/issues/70686

File: clang/lib/CodeGen/CGHLSLRuntime.cpp (hunk 1 above)
Comment (bogner) [reply clang/lib/CodeGen/CGHLSLRuntime.cpp:240]: I can't really find good documentation on what element types are really supposed to be allowed for resources, and dxc doesn't do a great job of diagnosing these things. However, it doesn't look like matrix types are supported here, at least in practice: https://hlsl.godbolt.org/z/aaze4ceYs

**Pull request merged**